man/man1/runbody4.1
.TH RUNBODY4 1NEMO "25 August 2025"

.SH "NAME"
runbody4 \- Hermite N-body code with optional stellar evolution

.SH "SYNOPSIS"
.nf
\fBrunbody4\fP [parameter=value]
\fBnbody4  \fP\fI < nbody4.in\fP
.fi

.SH "DESCRIPTION"
\fBrunbody4\fP is a frontend to Aarseth' NBODY4 program.
It creates a run directory, prepares
a parameter input file, and optionally input snapshot data, and runs
the NBODY4 program. This can be useful if you need to run a lot of
simulations unattended.
.PP
Output snapshots are by default not created, and must be  turned on
with a options flag (see \fBkz=\fP below). All of the input
parameter names are the same as their counterpart in
\fBnbody4\fP.
.PP
Note that this front-end is not designed to run every possible way
in which NBODY4 can be run, but it most useful to run initial conditions
from a snapshot file.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP
input file (optional, see \fBnbody=\fP to generate your own input 
data on the fly). If input data are given, it forces KZ(4)=2 (see 
below).  
Default: not used.
.TP
\fBoutdir=\fP
Output run directory. This directory should not exist, although
can be overriden with NEMO's standard \fBerror=\fP system keyword.
No default.
.TP
\fBnbody=\fP
Total particle number. The maximum value (NMAX) is compiled into
the NBODY4 program. Note that even if \fIin=\fP is supplied, this
parameter is needed.
No default.
.TP
\fBnfix=\fP
Output frequency of data save or binaries; KZ(3 & 6)
[Default: 1]
.TP
\fBnrand=\fP
Random number sequence skip.
[Default: 0]
.TP
\fBnrun=\fP
Run identification index.
[Default: 1]
.TP
\fBeta=\fP
Time-step parameter for force polynomials.
[Default: 0.02]
.TP
\fBdeltat=\fP
Output time interval in units of the crossing.
[Default: 0.25]
.TP
\fBtcrit=\fP
Termination time in units of the crossing time.
[Default: 5]
.TP
\fBqe=\fP
Energy tolerance (restart if DE/E > 5*QE & KZ(2) > 1)
[Default: 2e-5]
.TP
\fBeps=\fP
Potential softening parameter     
[Default: 0.01]
.TP
\fBkz=\fP
Non-zero options for alternative paths (see below) 
[Default: 1,2,0::4,1::2,0::2,1]
.nf
.ta +0.5i
1	COMMON save on unit 1 at end of run (=2: every 100*NMAX steps).
2	COMMON save on unit 2 at output (=1); restart if DE/E > 5*QE (=2).
3	Basic data on unit 3 at output (freq. NFIX; >1: cluster + tail).
4 #	Binary diagnostics on unit 4 (# threshold levels = KZ(4) < 10).
5	Initial conditions (#22 =0; =0: uniform & isotropic sphere;
      	      =1: Plummer; =2: two Plummer models in orbit, extra input;
      	      =3: massive perturber and planetesimal disk, extra input).
      	      =4: massive initial binary, extra input; output on unit 35).
6	Output of significant & regularized binaries (=1, 2, 3 & 4).
7	Lagrangian radii (>0: RSCALE; =2, 3, 4: output units 6, 7, 12;
      	      >=5: density & rms velocity at given radii on unit 26 & 27;
              >=5: average mass at given radii on unit 36 every DELTAT;
      	       =6: Lagrangian radii for two mass groups on unit 31 & 32.
8	Primordial binaries (=1 & >=3; >0: BINOUT; >2: BINDAT; >3: HIDAT).
9	Individual bodies printed at output time (MIN(5**KZ9,NTOT)).
10	Diagnostic KS output (>0: begin; >1: end; >=3: each step).
11 #	Synchronization of circular orbits (suppressed).
12	Disk shocks (=1: standard model) or interstellar clouds (< 0).
13	Scaling of time (1: variable by t_cr; 2: variable by t_r;
      	       -1: constant scaling to t_r; -2: constant scaling to t_c).
14	External force (=1: linearized; -1: cutoff; =2: point-mass galaxy;
      	   =3: point-mass + disk + logarithmic halo in any combination).
15	Triple, quad, chain (#30 > 0) or merger search (>1: full output).
16	Updating of regularization parameters (RMIN, DTMIN & ECLOSE).
17	Modification of ETA (>=1) & ETAU (>1) by tolerance QE.
18	Hierarchical systems (=1: diagnostics; =2: primordial; =3: both).
19	Stellar evolution and mass loss (=1: old supernova scheme;
      	            =3: Eggleton, Tout & Hurley; >4: Chernoff--Weinberg).
20	Initial mass function (=0,1: Salpeter; >1: various, see IMF).
                    -3=Eggleton w/ original typo    3=Eggleton book version
21	Extra output (>0: model, etc; >1: CENTRE; >2: MTRACE; >3: GLOBAL).
22	Initial conditions on unit 10 (=1: output; =2,3(unscaled): input).
23	Escaper removal (>1: diagnostics in file ESC; =2: angles unit #6;
      	                 >1: tidal tails output if #14 = 3).
24	Initial conditions for subsystem (routine SCALE; KZ(24) = #).
25 #	Partial reflection of KS binary orbit (GAMMA < GMIN; suppressed).
25	HR diagnostics of evolving stars (output of B & S on #82 & 83).
26	Slow-down of two-body motion (=1: KS binary; =2: chain binary).
27	Two-body interactions (-2: RADIUS = 0; -1: collision detection;
      	                       =1: sequential circ; > 0: collision).
28	(not used).
29 #	Boundary reflection for hot system (suppressed).
30	Chain regularization (=1: basic; >1: main output; >2: each step).
31	Centre of mass correction after energy check.
32	Increase of output intervals (based on single particle energy).
33	Block-step diagnostics at main output (=2: active pipes).
34 #	Roche lobe overflow (suppressed).
35	Time offset (global time from TTOT = TIME + DTOFF; offset = 100).
36	Step reduction for hierarchical systems (not recommended).
37	Step reduction for encounters with high-velocity particles.
38 #	Multiple use of GRAPE-6 (sleep 1 sec after each timer check).
39	Neighbour list (=-1: on host; =0: full list or closest on GRAPE).
40	(not used).

#	currently surpressed
.fi
.TP
\fBxtpar1=\fP
Mass of external Plummer model (KZ(15) = 1;
[Default: 1]
.TP
\fBxtpar2=\fP
Length scale for Plummer model (KZ(15) = 1)
[Default: 2]
.TP
\fBzmgas=\fP
Mass scale for logarithmic potential (KZ(15) = 2)
.TP
\fBrgas=\fP
Length scale for logarithmic potential (KZ(15) = 2)
.TP
\fBalphas=\fP
Power-law index for initial mass function (routine DATA)
[Default: 2.3]
.TP
\fBbody1=\fP
Maximum particle mass before scaling   
[Default: 5.0]
.TP
\fBbodyn=\fP
Minimum particle mass before scaling   
[Default: 1.0]
.TP
\fBq=\fP
Virial ratio (q=0.5 for virial equilibrium)  
[Default: 0]
.TP
\fBvxrot=\fP
XY-velocity scaling factor (> 0 for solid-body rotation)
[Default: 0]
.TP
\fBvzrot=\fP
Z-velocity scaling factor (not used if VXROT =
[Default: 0]
.TP
\fBrbar=\fP
Virial radius in pc (for scaling to physical
[Default: 1]
.TP
\fBzmbar=\fP
Mean mass in solar units   
[Default: 1]
.TP
\fBxcm=\fP
Displacement for subsystem (routine SCALE; KZ(17)) 
.TP
\fBecc=\fP
Eccentricity of relative motion for subsystem (ECC =< 1)
.TP
\fBKZ#=\fP
This is an indexed keyword,and the upper case version of the kz= array keyword. This
way the user can override single elements of the kz= vector, and using a default
for the long kz= vector. For example "kz=@kz.def KZ22=2 KZ5=2". See \fIgetparam(3NEMO)\fP
for a description of indexed keywords.  No default.

.SH "EXAMPLES"
\fBnbody4\fP can also be used to generate snapshots. Here is an example
to create a 1024 body homogeneous sphere in the directory
run1 and a plummer sphere in run2.
.nf
    % runbody4 "" run1 1024 tcrit=0 KZ5=0 KZ22=1
    % u3tos run1/OUT3 run1/run1.snap mode=4

    % runbody4 "" run2 1024 tcrit=0 KZ5=1 KZ22=1
    % u3tos run2/OUT3 run2/run2.snap mode=4

.fi
.PP
You can also supply existing NEMO snapshots as initial conditions
.nf
    % runbody4 run2/run2.snap run3 tcrit=1
.fi

.SH "BUGS"
Scaling can cause output to become out of bounds.
.PP
It appears to just create small systems, like nbody=10 tcrit=0, will hang the program.

.SH "SEE ALSO"
nbody4(1NEMO), nbody2(1NEMO), snapshot(5NEMO), u3tos(1NEMO), stou4(1NEMO), nbody4(5NEMO)

.SH "FILES"
.nf
.ta +2i
$NEMO/src/nbody/evolve/aarseth/tools	code
$outdir/fort.1                       	restart dump (compile time dep. size)
$outdir/fort.2                         	restart dump (compile time dep. size)
$outdir/fort.4                         	restart dump (compile time dep. size)
$outdir/fort.7                         	warning on nblist
$outdir/fort.8                         	T=
$outdir/fort.12				lagrangian radii?
$outdir/fort.16				ksrect
$outdir/fort.26				density
$outdir/fort.27				velocity
$outdir/fort.29				hivel
$outdir/fort.40				?
$outdir/fort.82				?
$outdir/fort.83				list of (several) times the N particles with something
$outdir/fort.88				T,N,...
$outdir/ESC				List of times and escapers with their properties

$outdir/OUT3                        	particle dump (see \fIu3tos(1NEMO)\fP)
HIDAT
OUT9

.SH "AUTHOR"
Peter Teuben

.SH "HISTORY"
.nf
.ta +1.25i +4.5i
28-feb-2006	V0.1 Created in Cambridge	PJT
10-mar-2006	V0.4 for a more formal version in NEMO, include Brut4's nbody4 version	PJT
13-mar-2006	V0.5 minor updated via Sverre (fixing small N initialization)	PJT/SJA
26-aug-2025	V0.5 fixed IMF equation typo info for KZ(20)	PJT/SAV
.fi


src/nbody/evolve/aarseth/tools/runbody4.c
/*
 *  MAIN.C: NEMO driver for nbody4
 *	
 *  28-feb-06    V0.1     Created, in Cambridge                          PJT
 *   8-mar-06    V0.3     clean up keywords, create Nbody on the fly     PJT
 *  10-mar-06    V0.4     unit10 now binary, parallel to nbody1,2
 *  17-sep-2013  V0.5     using new run interface                        PJT
 *                         
 *  TODO:   only a few special kz() cases were implemented. See nbody4.5 for the full set
 * 
 * VJ bench:       492", code halted tol exceeded at T=40  (3.2 GHz P4)
 * Code bench:    1448"                                                         2642" (laptop, P1.6GHz)
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <history.h>
#include <run.h>

#include <snapshot/snapshot.h>
#include <snapshot/body.h>
#include <snapshot/get_snap.c>

#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>

string defv[] = {
    "in=\n            input file (optional - see nbody= ) ",
    "outdir=???\n     output run directory (required)",

    "nbody=1000\n     Total particle number (<= NMAX).",
    "nfix=1\n         Output frequency of data save or binaries; KZ(3 & 6)",
    "ncrit=5\n        Final particle number (alternative termination criterion)",
    "nrand=50000\n    Random number sequence skip",
    "nrun=1\n         Run identification index",

    "eta=0.02\n       Time-step parameter for irregular force polynomial",
    "dtadj=2.0\n      check",
    "deltat=10.0\n    Output time interval in units of the crossing time",
    "tcrit=100.0\n    Termination time in units of the crossing time",
    "qe=2e-5\n        Energy tolerance (restart if DE/E > 5*QE & KZ(2) > 1)",
    "rbar=1.0\n       mean radius of system?",
    "zmbar=0.5\n      mean mass of system, in solar units",

#if 0
    "kz=0 1 0 0 1 0 1 0 0 0  0 0 0 1 1 1 0 0 3 4  1 0 2 0 0 1 0 0 0 1  0 0 0 0 0 0 1 0 0 0\n"
#else
    "kz=0 0 1 0 1 0 5 0 0 0  0 0 0 0 1 0 0 0 3 0  1 0 2 0 1 2 0 0 0 2  0 0 0 0 0 0 1 0 0 0\n"
#endif
      "     Non-zero options for alternative paths (see below)\n"
      "       1  COMMON save on unit 1 at end of run (=2: every 100*NMAX steps).\n"
      "       2  COMMON save on unit 2 at output (=1); restart if DE/E > 5*QE (=2).\n"
      "       3  Basic data on unit 3 at output (freq. NFIX; >1: cluster + tail).\n"
      "     # 4  Binary diagnostics on unit 4 (# threshold levels = KZ(4) < 10).\n"
      "       5  Initial conditions (#22 =0; =0: uniform & isotropic sphere;\n"
      "                =1: Plummer; =2: two Plummer models in orbit, extra input;\n"
      "                =3: massive perturber and planetesimal disk, extra input).\n"
      "                =4: massive initial binary, extra input; output on unit 35).\n"
      "       6  Output of significant & regularized binaries (=1, 2, 3 & 4).\n"
      "       7  Lagrangian radii (>0: RSCALE; =2, 3, 4: output units 6, 7, 12;\n"
      "                >=5: density & rms velocity at given radii on unit 26 & 27;\n"
      "                 =6: Lagrangian radii for two mass groups on unit 31 & 32.\n"
      "       8  Primordial binaries (=1 & >=3; >0: BINOUT; >2: BINDAT; >3: HIDAT).\n"
      "       9  Individual bodies printed at output time (MIN(5**KZ9,NTOT)).\n"
      "      10  Diagnostic KS output (>0: begin; >1: end; >=3: each step).\n"
      "    # 11  Synchronization of circular orbits (suppressed).\n"
      "      12  Disk shocks (=1: standard model) or interstellar clouds (< 0).\n"
      "      13  Scaling of time (1: variable by t_cr; 2: variable by t_r;\n"
      "                 -1: constant scaling to t_r; -2: constant scaling to t_c).\n"
      "      14  External force (=1: linearized; -1: cutoff; =2: point-mass galaxy;\n"
      "             =3: point-mass + disk + logarithmic halo in any combination).\n"
      "      15  Triple, quad, chain (#30 > 0) or merger search (>1: full output).\n"
      "      16  Updating of regularization parameters (RMIN, DTMIN & ECLOSE).\n"
      "      17  Modification of ETA (>=1) & ETAU (>1) by tolerance QE.\n"
      "      18  Hierarchical systems (=1: diagnostics; =2: primordial; =3: both).\n"
      "      19  Stellar evolution and mass loss (=1: old supernova scheme;\n"
      "                      =3: Eggleton, Tout & Hurley; >4: Chernoff--Weinberg).\n"
      "      20  Initial mass function (=0,1: Salpeter; >1: various, see IMF).\n"
      "      21  Extra output (>0: model, etc; >1: CENTRE; >2: MTRACE; >3: GLOBAL).\n"
      "      22  Initial conditions on unit 10 (=1: output; =2,3(unscaled): input).\n"
      "      23  Escaper removal (>1: diagnostics in file ESC; =2: angles unit #6;\n"
      "                           >1: tidal tails output if #14 = 3).\n"
      "      24  Initial conditions for subsystem (routine SCALE; KZ(24) = #).\n"
      "    # 25  Partial reflection of KS binary orbit (GAMMA < GMIN; suppressed).\n"
      "      25  HR diagnostics of evolving stars (output of B & S on #82 & 83).\n"
      "      26  Slow-down of two-body motion (=1: KS binary; =2: chain binary).\n"
      "      27  Two-body interactions (-2: RADIUS = 0; -1: collision detection;\n"
      "                                 =1: sequential circ; > 0: collision).\n"
      "      28  (not used).\n"
      "    # 29  Boundary reflection for hot system (suppressed).\n"
      "      30  Chain regularization (=1: basic; >1: main output; >2: each step).\n"
      "      31  Centre of mass correction after energy check.\n"
      "      32  Increase of output intervals (based on single particle energy).\n"
      "      33  Block-step diagnostics at main output (=2: active pipes).\n"
      "    # 34  Roche lobe overflow (suppressed).\n"
      "      35  Time offset (global time from TTOT = TIME + DTOFF; offset = 100).\n"
      "      36  Step reduction for hierarchical systems (not recommended).\n"
      "      37  Step reduction for encounters with high-velocity particles.\n"
      "    # 38  Multiple use of GRAPE-6 (sleep 1 sec after each timer check).\n"
      "      39  Neighbour list (=-1: on host; =0: full list or closest on GRAPE).\n"
      "      40  (not used).\n"
      "    # Currently surpressed",

    "dtmin=1e-5\n     time-step criterion for regularization search",
    "rmin=1e-4\n      distance criterion for regularization search",
    "etau=0.2\n       ",
    "eclose=1.0\n     ",
    "gmin=1e-6\n      ",
    "gmax=0.001\n     ",

    "alpha=2.3\n      Power-law index for initial mass function",
    "body1=10.0\n     Maximum particle mass before scaling",
    "bodyn=0.2\n      Minimum particle mass before scaling",
    "nbin0=0\n        number of primordial binaries",
    "zmet=0.02\n      ",
    "epoch0=0\n       epoch of star formation (in Myr)",
    "dtplot=10.0\n    plotting interval (nbody units)",

    "q=0.5\n          Virial ratio (q=0.5 for virial equilibrium)",

    "apo=6.0\n        kz(5) = 2 or 3",
    "ecc=0.5\n        Eccentricity of relative motion for subsystem (ECC =< 1)",
    "n2=500\n         kz(5) = 2",
    "scale=0.5\n      kz(5) = 2 or 3",
    "dmin=3\n         kz(5) = 3",
    "semi=1e-3\n      ",
    "m1=25.0\n        ",
    "m2=25.0\n        ",
    "ratio=0.0\n      ",
    "range=100.0\n    ",
    
    "kstart=1\n       Running mode (1=new 2=restart 3,4,5=restart w/ new par",
    "tcomp=10.0\n     Maximum allowed running time (minutes) **ignored **",
    "gpid=0\n         Number of grape boards (starting from 0) ** if applicable **",

    "format=%g\n      Format used for fort.10 input conditions if in= used",
    "KZ#=\n           [indexed] Override some kz= keywords",

    "VERSION=0.7\n    8-jan-2024 PJT",
    NULL,
};

string usage="Hermite N-body code";

#define KZ_MAX  40

#define KZ_OUT   3
#define KZ_INI   4
#define KZ_EXT  15
#define KZ_MER  17
#define KZ_COM  18

void nemo_main(void)
{
  int nbody=0, nfix, ncrit, nrand, nrun, kstart, gpid, nbin0, n2;
  real eta, dtadj, deltat, tcrit, qe, tcomp;
  int k, nkz, kz[KZ_MAX];
  int  bits;
  real dtmin,rmin,etau,eclose,gmin,gmax, tsnap;
  real alpha, body1, bodyn, zmet, epoch0, dtplot;
  real q, rbar, zmbar;
  real apo, ecc, dmin, semi, scale, m1, m2, ratio, range;
  Body *bp, *btab = NULL;

  string exefile = "nbody4";
  string parfile = "nbody4.in";
  string rundir = getparam("outdir");
  string fmt = getparam("format");
  char dname[256], runcmd[256], fmt7[256];
  stream datstr, histr, instr, outstr;

  kstart = getiparam("kstart");
  tcomp =  getdparam("tcomp");
  gpid = getdparam("gpid");
  
  nfix = getiparam("nfix");
  ncrit = getiparam("ncrit");
  nrand = getiparam("nrand");
  nrun = getiparam("nrun");
  
  eta = getdparam("eta");
  dtadj = getdparam("dtadj");
  deltat = getdparam("deltat");
  tcrit = getdparam("tcrit");
  qe = getdparam("qe");
  rbar =  getdparam("rbar");
  zmbar =  getdparam("zmbar");
  
  nkz = nemoinpi(getparam("kz"),kz,KZ_MAX);
  if (nkz < 0) error("%d Syntax error kz=%s",nkz,getparam("kz"));
  for (k=nkz; k<KZ_MAX; k++) kz[k]=0;
  for (k=0; k<KZ_MAX; k++) dprintf(1,"%d ",kz[k]);
  dprintf(1,"\n");

  for (k=0; k<KZ_MAX; k++) {
    if (indexparam("KZ",k+1)) {
      dprintf(0,"KZ %d=%d\n",k+1,getiparam_idx("KZ",k+1));
	kz[k] = getiparam_idx("KZ",k+1);
    }
  }

  if (hasvalue("in")) {
    /* read snapshot and grab nbody.... */

    instr = stropen(getparam("in"),"r");
    get_history(instr);
    if (!get_tag_ok(instr, SnapShotTag)) error("no snapshot");
    get_snap(instr, &btab, &nbody, &tsnap, &bits);
    strclose(instr);
    if (kz[21] == 0) {
      warning("patching kz[22]=1 since in= was given");
      kz[21] = 2;
    }
  } else
    nbody = getiparam("nbody");


  dtmin = getdparam("dtmin");
  rmin = getdparam("rmin");
  etau = getdparam("etau");
  eclose= getdparam("eclose");
  gmin = getdparam("gmin");
  gmax = getdparam("gmax");

  
  alpha = getdparam("alpha");
  body1 = getdparam("body1");
  bodyn = getdparam("bodyn");
  nbin0 = getiparam("nbin0");
  zmet = getdparam("zmet");
  epoch0 = getiparam("epoch0");
  dtplot = getiparam("dtplot");

  q = getdparam("q");

  /* optional things, depending on kz5 */
  apo = getdparam("apo");
  ecc = getdparam("ecc");
  n2 = getiparam("n2");
  dmin = getdparam("dmin");
  scale = getdparam("scale");
  semi = getdparam("semi");
  m1 = getdparam("m1");
  m2 = getdparam("m2");
  ratio = getdparam("ratio");
  range = getdparam("range");

  /* there are more optional kz-dependant keywords...
     see nbody4.5, they have not been encoded here
     by lack of Cambridge time 
  */

  run_mkdir(rundir);

  sprintf(dname,"%s/%s",rundir,parfile);
  datstr = stropen(dname,"w");    
  
    /*  New Run */
  
  if (kstart == 1) {
    
    fprintf(datstr,"%d %g %d\n",kstart,tcomp,gpid);
    fprintf(datstr,"%d %d %d %d %d\n",nbody,nfix,ncrit,nrand,nrun);
    fprintf(datstr,"%g %g %g %g %g %g %g\n",eta,dtadj,deltat,tcrit,qe,rbar,zmbar);

    for (k=0; k<KZ_MAX; k++) {
      fprintf(datstr,"%d ",kz[k]);
      if (k%10 == 9)     fprintf(datstr,"\n");
    }

    fprintf(datstr,"%g %g %g %g %g %g\n",dtmin,rmin,etau,eclose,gmin,gmax);
    fprintf(datstr,"%g %g %g %d %g %g %g\n",alpha,body1,bodyn,nbin0,zmet,epoch0,dtplot);

    if (kz[4] == 2)
      fprintf(datstr,"%g %g %d %g\n",apo,ecc,n2,scale);
    else if (kz[4] == 3) 
      fprintf(datstr,"%g %g %g %g\n",apo,ecc,dmin,scale);
    else if (kz[4] == 4) 
      fprintf(datstr,"%g %g %g %g\n",semi,ecc,m1,m2);

    fprintf(datstr,"%g 0  0 0 \n",q);

    if (kz[7] == 1 || kz[7] == 3)
      fprintf(datstr,"%g %g %g %g 0 0 0 \n",semi,ecc,ratio,range);




    strclose(datstr);

    run_cd(rundir);


    histr = stropen("history","w");
    put_history(histr);
    strclose(histr);

    /*
     * two options here: the snapshot can be written in ascii, and read
     * or stou4 can be used, thereby preserving more digits reliably....
     * in some future version this will be aligned with nbody1 and 2
     * or vice versa....
     */

    if (hasvalue("in")) {
#if 0
      string infile;
      dprintf(0,"Using stou4 to convert data for nbody4\n");
      infile = getparam("in");
      if (*infile == '-') {           /* do something special for pipes */
	sprintf(runcmd,"stou4 %s fort.10",infile);
      } else if (*infile == '/') {    /* regular file */
	sprintf(runcmd,"stou4 %s fort.10",infile);
      } else
	error("in=%s must be an absolute pathname or special file",infile);
      dprintf(0,"%s\n",runcmd);
      if (run_sh(runcmd)) error("Error converting input data");
#else
      dprintf(0,"Using ascii printout with format=%s to convert data for nbody4\n",fmt);
      sprintf(fmt7,"%s %s %s %s %s %s %s\n",fmt,fmt,fmt,fmt,fmt,fmt,fmt);
      outstr = stropen("fort.10","w");
      for (bp=btab; bp<btab+nbody; bp++)
	fprintf(outstr,fmt7,
		Mass(bp),
		Pos(bp)[0],Pos(bp)[1],Pos(bp)[2],
		Vel(bp)[0],Vel(bp)[1],Vel(bp)[2]);
      strclose(outstr);
#endif
    }

    sprintf(runcmd,"%s < %s",exefile,parfile);
    run_sh(runcmd);

    sprintf(runcmd,"u3tos in=OUT3 out=OUT3.snap mode=4");
    run_sh(runcmd);
  } else {
    error("kstart=%d not yet supported for NBODY4",kstart);
  }
}
